cnqr/_src/convex.py

```diff
@@ -0,0 +1,207 @@
+  dim_data: int = 2
+  dim_context: int = 1
+  context_dense_layer: str = nn.Dense  # TODO: try with BjorckDense
+  gaussian_map: Tuple[jnp.ndarray, jnp.ndarray] = None
+
+  def setup(self):
+    self.num_hidden = len(self.dim_hidden)
+    rescale = PositiveDense.inv_rectifier_fn
+    self.use_init = False
+    factor, mean = self._compute_identity_map(self.dim_data)  # for the computation of potential q
+
+    # ===== Input encoding =====
+    # TODO: phi => encoding the input of context network (upscale to dim_hidden[0])
+    phi = nn.Dense(
+      self.dim_hidden[0],
+      name=f'phi_dense',
+      kernel_init=self.init_fn(),
+      bias_init=nn.initializers.constant(0.0),
+      use_bias=True,
+    )
+    self.phi = phi
+    # => Simple linear ? Or upsampling ?
+
+    # encoding the input of convex network (potential that will need to be upscaled to dum_hidden[0])
+    q = PosDefPotentials(
+      self.dim_data,
+      name='q_potential',
+      num_potentials=1,
+      kernel_init=lambda *args, **kwargs: factor,
+      bias_init=lambda *args, **kwargs: mean,
+      use_bias=True,
+    )
+    self.q = q
+
+    # ===== Context network =====
+    # (now simple dense, then will be replaced with lipschitz)
+    f_x = []  # dense layers (with bias), init: weights -> identity, bias -> 0
+    for i in range(0, self.num_hidden + 1):
+      output_dim = 1 if i == self.num_hidden else self.dim_hidden[i]
+      f_x.append(
+          self.context_dense_layer(
+              output_dim,
+              name=f'f_x_dense_{i}',
+              kernel_init=self._identity_initializer,
+              bias_init=nn.initializers.constant(0.0),
+              use_bias=True,
+          )
+      )
+    self.f_x = f_x
+
+    # ===== Context modulators =====
+    f_u = []  # dense layers (with bias), init: weights -> approx(0), bias -> 1
+    for i in range(0, self.num_hidden + 1):
+      output_dim = 1 if i == self.num_hidden else self.dim_hidden[i]
+      f_u.append(
+          self.context_dense_layer(
+              output_dim,
+              name=f'f_u_dense_{i}',
+              kernel_init=self.init_fn(self.init_std),
+              bias_init=nn.initializers.constant(1.0),
+              use_bias=True,
+          )
+      )
+    self.f_u = f_u
+
+    f_v = []  # dense layers (with bias), init: weights -> approx(0), bias -> 0
+    for i in range(0, self.num_hidden + 1):
+      output_dim = 1 if i == self.num_hidden else self.dim_data
+      f_v.append(
+          self.context_dense_layer(
+              output_dim,
+              name=f'f_v_dense_{i}',
+              kernel_init=self.init_fn(self.init_std),
+              bias_init=nn.initializers.constant(0.0),
+              use_bias=True,
+          )
+      )
+    self.f_v = f_v
+
+    f_w = []  # dense layers (with bias), init: weights -> approx(0), bias -> 0
+    for i in range(0, self.num_hidden + 1):
+      output_dim = 1 if i == self.num_hidden else self.dim_hidden[i]
```

Comment:
There is code duplication for `f_u` and `f_w`, and even for `f_v` in some extent.  
I suggest you create lists `output_dims_uw = self.dim_hidden + [1]` and `output_dims_v = self.dim_hidden + [self.dim_data]`.
Then you can replace `for i in range(0, self.num_hidden + 1):` by `for i, output_dim in enumerate(output_dims_uw):` (for example).